helix-term/src/commands.rs

```diff
@@ -3161,60 +3162,72 @@ pub mod insert {
+            });
 
-        if let Some(lsp::CompletionOptions {
-            trigger_characters: Some(triggers),
-            ..
-        }) = &capabilities.completion_provider
-        {
-            // TODO: what if trigger is multiple chars long
-            if triggers.iter().any(|trigger| trigger.contains(ch)) {
-                cx.editor.clear_idle_timer();
-                super::completion(cx);
-            }
+        if trigger_completion {
+            cx.editor.clear_idle_timer();
+            super::completion(cx);
         }
     }
 
     fn signature_help(cx: &mut Context, ch: char) {
+        use futures_util::FutureExt;
         use helix_lsp::lsp;
         // if ch matches signature_help char, trigger
-        let doc = doc_mut!(cx.editor);
-        // The language_server!() macro is not used here since it will
-        // print an "LSP not active for current buffer" message on
-        // every keypress.
-        let language_server = match doc.language_server() {
-            Some(language_server) => language_server,
-            None => return,
-        };
-
-        let capabilities = language_server.capabilities();
+        let (view, doc) = current!(cx.editor);
+        // lsp doesn't tell us when to close the signature help, so we request
+        // the help information again after common close triggers which should
+        // return None, which in turn closes the popup.
+        let close_triggers = &[')', ';', '.'];
+        // TODO support multiple language servers (not just the first that is found)
+        let future = doc
+            .language_servers_with_feature(LanguageServerFeature::SignatureHelp)
+            .iter()
+            .find_map(|ls| {
+                let capabilities = ls.capabilities();
+
+                match capabilities {
+                    lsp::ServerCapabilities {
+                        signature_help_provider:
+                            Some(lsp::SignatureHelpOptions {
+                                trigger_characters: Some(triggers),
+                                // TODO: retrigger_characters
+                                ..
+                            }),
+                        ..
+                    } if triggers.iter().any(|trigger| trigger.contains(ch))
+                        || close_triggers.contains(&ch) =>
+                    {
+                        let pos = doc.position(view.id, ls.offset_encoding());
+                        ls.text_document_signature_help(doc.identifier(), pos, None)
+                    }
+                    _ if close_triggers.contains(&ch) => ls.text_document_signature_help(
+                        doc.identifier(),
+                        doc.position(view.id, ls.offset_encoding()),
+                        None,
+                    ),
+                    // TODO: what if trigger is multiple chars long
+                    _ => None,
+                }
+            });
 
```

Comment:
it would make the diff a lot smaller and easier to review if you didn't move stuff like this into the `find_map` instead you can just use the generic `any` pattern @dead10ck proposed above to find the first LS that supports the feature and then keep the rest of the code unchanged

Reply:
Sorry about that.

I don't really see how the `any` pattern would be of help here. It's a little bit more complicated, that's why I have chosen a `find_map` for all requests that use the first supported language server:

* First find language servers by `LanguageServerFeature`
* Then filter them, based on the actual LSP capabilities, which is provided by all the requests like `goto_definition`, which returns `None`, if the language server doesn't support that request, and the first `Some(future)` is the actual request.

Maybe I've overlooked something? I don't currently see a more concise less diff-producing way to do this properly. Only by checking the LSP capabilities separately with something like `doc.language_servers_with_feature(<feature>).find(<check_capabilities>)`.
But this would either result in a double check of the capabilities (as they are checked in the client request methods) or requires a bigger refactor, AFAIK.

Reply:
I have been thinking about this a bunch more and I do think it's better to just check capabilities twice. Compared to all the moved code this is just less work to review and a single capability checking function.

There is also basically no runtime overhead to checking capabilities twice (you are just reading a field) whereas running the request in `find_map` can end-up executing quite a bit of complex code multiple times.

Thinking about it more I also think it's fine to associate config feature to a single capability. I can't really think of a use-case where this doesn't apply.

Reply:
Should I add this feature -> capability association in this PR and include it in `doc.language_servers_with_feature`, so that we can just unwrap the requests/futures?

Reply:
Refactored/Resolved in https://github.com/helix-editor/helix/commit/25080f61c5e6dadab07d0d2b193a6138fd7b1fd0